app/services/solr_mapper.rb

```diff
@@ -14,15 +14,90 @@ def initialize(metadata:, dataset_record_id:, dataset_record_set_id:)
   end
 
   # @return [Hash] the Solr document
+  # rubocop:disable Metrics/AbcSize
   def call
     {
       id: dataset_record_id,
-      dataset_record_set_id: dataset_record_set_id,
-      title: metadata.dig(:titles, 0, :title)
-    }
+      dataset_record_set_id_ss: dataset_record_set_id,
+      access_ssi: metadata['access'],
+      provider_ssi: metadata['provider'],
+      creators_struct_ss: metadata['creators'].to_json,
+      descriptions_tsim: retrieve_descriptions(metadata['descriptions'] || []),
+      doi_ssi: retrieve_dois(metadata['identifiers']),
+      provider_identifier_ssim: map_provider_identifiers(metadata['identifiers'], metadata['provider'])
+    }.merge(map_titles(metadata['titles']))
   end
+  # rubocop:enable Metrics/AbcSize
 
   private
 
   attr_reader :metadata, :dataset_record_id, :dataset_record_set_id
+
+  # Given titles from metadata, return field value based on title type
+  def title_fields(title_type, titles)
+    titles.select { |title_obj| title_obj.key?('title_type') && title_obj['title_type'] == title_type }.pluck('title')
+  end
+
+  def map_titles(titles_metadata)
+    titles = titles_metadata.reject { |title_obj| title_obj.key?('title_type') }.pluck('title')
+    subtitles = title_fields('Subtitle', titles_metadata)
+    alternative_titles = title_fields('AlternativeTitle', titles_metadata)
+    translated_titles =  title_fields('TranslatedTitle', titles_metadata)
+    other_titles = title_fields('Other', titles_metadata)
+
+    titles = assign_titles(titles, subtitles, alternative_titles, translated_titles, other_titles)
+
+    map_typed_title_fields(titles, subtitles, alternative_titles, translated_titles, other_titles)
+  end
+
+  def assign_titles(titles, subtitles, alternative_titles, translated_titles, other_titles)
+    # One of the title elements is required, but we will use one of the other titles when present
+    return titles unless titles.empty?
+
+    if !alternative_titles.empty?
+      alternative_titles
+    elsif !subtitles.empty?
+      subtitles
+    elsif !other_titles.empty?
+      other_titles
+    else
+      translated_titles
+    end
+  end
+
+  def map_typed_title_fields(titles, subtitles, alternative_titles, translated_titles, other_titles)
+    mapped_titles = {
+      title_tsim: titles
+    }
+    mapped_titles[:subtitle_tsim] = subtitles unless subtitles.empty?
```

Comment:
Can you use `if subtitles.present?` instead of `unless`?

Reply:
Yes

Reply:
I ended up removing .present? altogether b/c Solr indexing will handle empty arrays by just not adding them. so we should be good